### ISSUE TITLE
Incorrect prefix key

### ISSUE THREAD

**Issue by WimVDK** [akeeba/loginguard#107]
Please read the README.md in the repository's root and the support resources before submitting an issue.

### Steps to reproduce the issue
Relative fresh dev install with template and a few extensions added, akeeba loginguard didn't remember my preference to "remember me" after logging in. The next visit it would prompt for 2FA again. So wiped everything, installed fresh, install dev release pkg_loginguard-rev3694B18 (which fixes the backup codes not regenerating, see my previous solved issue) to get the below issue.

### Expected result

normal install

### Actual result

install failed with 

```
Warning
JInstaller: :Install: Error SQL Incorrect prefix key; the used key part isn't a string, the used length is longer than the key part, or the storage engine doesn't support unique prefix keys
Extension Install: SQL error processing query: DB function failed with error number 1089
Incorrect prefix key; the used key part isn't a string, the used length is longer than the key part, or the storage engine doesn't support unique prefix keys
SQL =

CREATE TABLE IF NOT EXISTS `#__loginguard_tfa`
(
    `id`         SERIAL,
    `user_id`    BIGINT UNSIGNED NOT NULL,
    `title`      VARCHAR(255)    NOT NULL,
    `method`     VARCHAR(100)    NOT NULL,
    `default`    TINYINT(1)      NOT NULL DEFAULT 0,
    `options`    LONGTEXT        null,
    `created_on` DATETIME        NULL,
    `last_used`  DATETIME        NULL,
    INDEX idx_user_id (`user_id`(100))
) ENGINE = InnoDB
  DEFAULT CHARSET = utf8mb4
  DEFAULT COLLATE = utf8mb4_unicode_ci;

Package Install: There was an error installing an extension: com_loginguard.zip
```

### Troubleshooting already performed

none yet, awaiting feedback

#### Mandatory information

Issues without this information can not and _will not_ be replied to.

* **Joomla! version**: 3.9.28
* **PHP version**: 7.4.21
* **Am I using HTTPS**: Yes
* **Server Operating System**: Linux
* **My device's Operating System**: Win 10 Pro x64
* **My device's Operating System version**: 21H1 b19043.1083
* **Browser**: Chrome, Firefox
* **Browser version**: Main browser Firefox 89.0.2

#### Good to have information

You can skip some or all of this information. However, the more information you provide the faster and better we can help.

* **MySQL version**: 5.5.5-10.2.39-MariaDB
* **Server type**: Apache
* **Hosting company / Local server**: self-host
* **Server Operating System version**: CentOS

### Additional comments

Akeeba Backup Pro, Admin tools Pro both install and work just fine.

**Comment by WimVDK** [comment]
I got notified you closed this. Want me to try a new dev release?

**Comment by nikosdion** [comment]
We have no childcare this week on a few hours’ notice. When I’ll be able to make a dev release is currently unknown.

**Comment by WimVDK** [comment]
No worries. Since only 1 file got changed, I'll find the file in the current dev release and copy/paste the updated code and make a new zip. That way it's the same as a new dev release.I'll let you know the result.On Jul 12, 2021 08:02, "Nicholas K. Dionysopoulos" ***@***.***> wrote:

We have no childcare this week on a few hours’ notice. When I’ll be able to make a dev release is currently unknown.

**Comment by nikosdion** [comment]
Please note that I’ve rewritten LoginGuard from the ground up. The previous dev release IIRC was from the old version. You won’t be able to find the file there.

Hold on until power is back so I can publish a dev release. Yeah, to compound an already fraught situation the power went out as well.

**Comment by nikosdion** [comment]
Developer's release: https://www.akeeba.com/download/developer-releases/loginguard-dev/revff94aca.html

**Comment by WimVDK** [comment]
I knew you were rewriting LG using the Joomla MVC from the previous issue :) That said, damn a power outage is never fun. (File was present in dev release pkg_loginguard-rev3694B18 though. Just had to open up the com_loginguard zip inside the package zip to find it. But didn't make the change since you had already published the new dev build.

Download and install attepted. Fails to install entirely with the below message.

![afbeelding](https://user-images.githubusercontent.com/14872315/125314059-25fcdd80-e336-11eb-94d2-872229878745.png)

appears, rendering the entire site non functional. Had to go into the database and set the status for
![afbeelding](https://user-images.githubusercontent.com/14872315/125314699-b76c4f80-e336-11eb-8a87-cb99423801c8.png)
 colum "enabled" to 0. This rendered the site back operational and the install screen gave the "installation succesful" message.

**Comment by nikosdion** [comment]
I'll check the action log plugin.

Meanwhile you can delete the `plugins/actionlog/loginguard` directory.

**Comment by WimVDK** [comment]
change the db entry back, leave as is or remove it as well? If i can be of any further assistance don't hesitate to ask. Thanks for being so quick on the ball.

**Comment by nikosdion** [comment]
No further change, just remove that folder to regain access to the site.

**Comment by WimVDK** [comment]
just a question then. is making a db change needed then in such a case, if removing a folder has the same effect. Just trying to learn here.

**Comment by nikosdion** [comment]
New dev release: https://www.akeeba.com/download/developer-releases/loginguard-dev/reve4cf86c.html

Regarding your question: if you remove a plugin's folder the plugin can no longer be found and Joomla does not load it. You should NOT mess with the database, though. You can't just remove the `#__extensions` record for the plugin. All records on that table are assets which means they have a corresponding entry in the `#__assets` table. The assets table is a tree which means that if you delete a record without updating the `lft` and `rgt` values you are creating a hole in it. The hole can cause problems with ACLs further down the road. When you install or uninstall an extension Joomla uses its tree model code to change the left/right values of all records, making sure there are no holes.

There are countless pitfalls like that when you are doing database surgery. That's why I spend a lot of time reading Joomla's code, experimenting with different problems on sacrificial development sites and fain the experience to come up with the best instructions which are a. clear and easy to follow and b. won't screw up your site. As with all things technical, the devil's in the details. If you want to save yourself pain follow the instructions of developers without improvising or mixing up the order of steps. Even when we tell you to do something that on the face of it is unrelated, strange, or outright stupid there's a good reason which typically requires a five pages long technical explanation. If you want to see an example of that look no further than https://www.akeeba.com/news/1746-working-around-joomla-s-broken-extensions-updater.html where I explain the problems we have experienced with Joomla's extensions installer / updater and their technical explanation. A lot of devils hiding in a lot of details...

**Comment by WimVDK** [comment]
Uninstalled previous dev release, to be sure. This new one installed perfectly. Curious if this one will remember the "remember me" status, previous 2 builds still asked for a 2FA code after closing browser for 30min and accessing the site again. (Unless i'm missunderstanding this functionality)

**Comment by nikosdion** [comment]
The “remember me” feature generates a unique browser fingerprint and stores it in an HTTP-only cookie (with some anti-tampering, of course) when you complete Two Step Verification. The next time you try to run a 2SV the same fingerprint is generated. If it matches the stored one the Two Step Verification step is skipped over.

If your browser and its extensions are blocking tracking very aggressively the two fingerprints will not match, therefore you will be asked to provide Two Step Verification again.

Simply put, the more paranoid you are about preventing tracking, the less likely it is that Remember Me will work. At the same time, the more paranoid you are about preventing tracking the more security-conscious you are, therefore you wouldn't mind the additional security prompts. That's the premise upon which this feature was built.

**Comment by WimVDK** [comment]
My main browser is Firefox, i do remember firefox becoming more anti-cookie focussed. My current setting is on "modified", i'll try default and see how that works out. I'll keep you posted!

**Comment by WimVDK** [comment]
Nicholas, forgive me for reactivating this but i did some more digging. While i have no explanation for any of it. I compared 4 sites that were both running loginguard while my browser settings remained the same. I could not immediately find any explanation as to why 3 of them worked fine with the "trusted silent fingerprint (after entering login credentials)" and this most recent developed website just would keep asking for 2FA code. Until i stumbled upon something

3/4 were using the latest stable release of 4.0.0, while the "non functional silent fingerprint" one was now running dev release revE4CF86C (the latest one you provided in this issue) and i was getting prompted to update to version 4.0.0. I figured to take the gamble and see if that would restore silent 2FA authorization. 4.0.0 installed right over the top of the latest dev release, but during the update process it did ask me for 2FA authorization (code, don't know if this is normal, previous over the top ones didn't ask for a 2FA code). I entered the code and update completed succesfully.

I then proceeded to closing the browser entirely (while making no changes to the cookie & privacy settings currently active) and waited an hour. A time period i know from previous testing would require me to enter a 2FA code again on the previous dev releases. I logged back in with user and password and was immediately forwarded to the administration panel where i saw the notice "<user> had a succesful Two Step Verification with method "Silent Verification (Trusted Browser Fingerprint)", so i know that updating/downgrading from dev release revE4CF86C to release 4.0.0 restores silent login.

The downside of course is that it breaks the regeneration of backup codes and "restores" [issue 106](https://github.com/akeeba/loginguard/issues/106).

That leads me that SOMETHING (i dont know what) must have changed in the internal code that causes the trusted browser fingerprint to "break" between 4.0.0 and the current dev releases (before or after fixing issue 106). I hope this info finds you well and gives some info as to where to look.

As always i appreciate your time and assistance in this matter and am always willing to test potential fixes.

**Comment by nikosdion** [comment]
Try the latest dev release (https://www.akeeba.com/download/developer-releases/loginguard-dev.html). The previous dev release was not setting the cookie at all.

A word of caution about cookie banner plugins. If the user has not accepted cookies yet, the cookie banner plugin will silently unset all cookies on the page except the session cookies. This includes Joomla's Remember Me _and_ LoginGuard's Remember Me which uses the same cookie method (down to cookie series and token) as Joomla's Remember Me. I spotted this very obvious (in retrospect!) issue when I was doing some local tests.

**Comment by WimVDK** [comment]
Installed. Will report back tomorrow or friday. Bedtime now (bbq tomorrow ^^). Testing site has no cookies on the front-end (no login fuctionality at all) only a session cookie that doesn't seem to get blocked despite the extension (not ALG) being set to explicitly block that too. But that's an issue i need to raise with them. 

It's [this](https://storejextensions.org/extensions/gdpr.html) extension i use for all my GDPR managing, perhaps you know of it. It's set to "Preventively block cookies" and remember me worked on 4.0.0. If it "bugs" out again you'll be the first to know!

**Comment by WimVDK** [comment]
sorry for the delay Nicholas. I had an unexpected funeral to deal with these last 2 weeks. My last remaining grandparent had passed away at the respected age of 95. So yeah, that was no fun to deal with. Anyway i have installed release rev16D1DB0, logged in, used 2FA, logged out. Logged back in and i was NOT asked for 2FA code again. Instead the "trusted fingerprint" appeared up top on 2nd login so i expect everything is fine.

I see that you published v5.0.0 so i'll test that one too.

**Comment by nikosdion** [comment]
I'm sorry for your loss.

Best try 5.0.1 I released today.

Sorry for the long delay in my reply too. Between evacuating because of the fires and running like a headless chicken to finish up releases before Joomla 4 stable I had barely any time to reply here.

**Comment by WimVDK** [comment]
Thanks, but don't worry about it. Yeah those wildfires have been all over the news here in Belgium. Luckily there's neighbouring countries coming to assist. Hopefully you, your family and possessions are safe (or not too much lost).

**Comment by nikosdion** [comment]
The fire didn't reach us (well, my mum's house where we were staying for the summer), it stopped one village over. Nothing lost, thankfully. We do have fire insurance for the house after a wildfire in '07 stopped literally at the edge of the property. In case you're wondering, house insurance is optional here. No, I'm not sure how that makes sense either!

**Comment by WimVDK** [comment]
![afbeelding](https://user-images.githubusercontent.com/14872315/129603705-9e6dd1a4-fff2-4889-8baf-70495ee38103.png)

Can i assume based on the timestamps, everything is working as it should? 501 is installed. To be sure i checked again if the backup code reset still worked and it does. Pretty sure that for the moment that fixes everything known to date.

**Comment by nikosdion** [comment]
Yes, 5.0.1 is the latest version at the time of this writing
